File: VB/App_Code/CustomFileSystemProvider.vb
Option Infer On

Imports DevExpress.Web
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.IO
Imports System.Linq
Imports System.Security.Principal
Imports System.Web
Imports System.Xml.Linq

Public Class CustomFileSystemProvider
    Inherits PhysicalFileSystemProvider

    Public Sub New(ByVal rootFolder As String)
        MyBase.New(rootFolder)
    End Sub
    Protected Friend Overridable Function MapPath(ByVal virtualPath As String) As String
        Return System.Web.Hosting.HostingEnvironment.MapPath(virtualPath)
    End Function

    Private extensionsDisplayName_Renamed As Dictionary(Of String, String)
    Private ReadOnly Property ExtensionsDisplayName() As Dictionary(Of String, String)
        Get
            If extensionsDisplayName_Renamed Is Nothing Then
                extensionsDisplayName_Renamed = XDocument.Load(MapPath("~/Content/ExtensionsDisplayName.xml")).Descendants("Extension").ToDictionary(Function(n) n.Attribute("Extension").Value, Function(n) n.Attribute("DisplayName").Value)
            End If
            Return extensionsDisplayName_Renamed
        End Get
    End Property
    Public Overrides Iterator Function GetFiles(ByVal folder As FileManagerFolder) As IEnumerable(Of FileManagerFile)
        For Each file As FileManagerFile In MyBase.GetFiles(folder)
            Dim fileInfo = New FileInfo(MapPath(file.FullName))
            Dim fileProperties As New FileManagerFileProperties() With {.Metadata = New Dictionary(Of String, Object)()}
            Dim fileSecurity = System.IO.File.GetAccessControl(MapPath(file.FullName))
            Dim sid = fileSecurity.GetOwner(GetType(SecurityIdentifier))
            Dim myFileVersionInfo As FileVersionInfo = FileVersionInfo.GetVersionInfo(MapPath(file.FullName))
            Dim version = myFileVersionInfo.FileVersion
            Dim description As String = myFileVersionInfo.FileDescription
            fileProperties.Metadata.Add("Name", file.Name)
            fileProperties.Metadata.Add("Type", ExtensionsDisplayName(file.Extension))
            fileProperties.Metadata.Add("CreationDate", fileInfo.CreationTime.ToString("U"))
            fileProperties.Metadata.Add("AccessedDate", fileInfo.LastAccessTime.ToString("U"))
            fileProperties.Metadata.Add("Attributes", fileInfo.Attributes)
            fileProperties.Metadata.Add("LastWriteTime", fileInfo.LastWriteTime.ToString("U"))
            fileProperties.Metadata.Add("Owner", sid.Value)
            fileProperties.Metadata.Add("Description", description)
            fileProperties.Metadata.Add("Size", fileInfo.Length.ToString("#,#") & " bytes")
            Yield New FileManagerFile(Me, folder, file.Name, file.Id, fileProperties)
        Next file
    End Function
    Public Overrides Iterator Function GetFolders(ByVal parentFolder As FileManagerFolder) As IEnumerable(Of FileManagerFolder)
        For Each folder As FileManagerFolder In MyBase.GetFolders(parentFolder)
            Dim folderInfo = New FileInfo(MapPath(folder.FullName))
            Dim folderProperties As New FileManagerFolderProperties() With {.Metadata = New Dictionary(Of String, Object)()}
            Dim folderSecurity = File.GetAccessControl(MapPath(folder.FullName))
            Dim sid = folderSecurity.GetOwner(GetType(SecurityIdentifier))
            folderProperties.Metadata.Add("Name", folder.Name)
            folderProperties.Metadata.Add("Type", "Folder")
            folderProperties.Metadata.Add("CreationDate", folderInfo.CreationTime.ToString("U"))
            folderProperties.Metadata.Add("AccessedDate", folderInfo.LastAccessTime.ToString("U"))
            folderProperties.Metadata.Add("Attributes", folderInfo.Attributes)
            folderProperties.Metadata.Add("LastWriteTime", folderInfo.LastWriteTime.ToString("U"))
            folderProperties.Metadata.Add("Owner", sid.Value)
            Dim directoryInfo = New DirectoryInfo(MapPath(folder.FullName))
            Dim filesInfo = directoryInfo.GetFiles("*", SearchOption.AllDirectories)
            Dim folderSize As Long = 0
            For Each info In filesInfo
                folderSize += info.Length
            Next info
            Dim filesCount = filesInfo.Length
            Dim subDirectoriesCount = directoryInfo.GetDirectories("*", SearchOption.AllDirectories).Length
            folderProperties.Metadata.Add("Size", folderSize.ToString("#,#") & " bytes")
            folderProperties.Metadata.Add("Contains", String.Format("Files: {0}, Folders: {1}", filesCount, subDirectoriesCount))
            Yield New FileManagerFolder(Me, folder, folder.Name, folder.Id, folderProperties)
        Next folder
    End Function
End Class

File: CS/App_Code/CustomFileSystemProvider.cs
using DevExpress.Web;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Security.Principal;
using System.Web;
using System.Xml.Linq;

public class CustomFileSystemProvider : PhysicalFileSystemProvider {
    public CustomFileSystemProvider(string rootFolder) : base(rootFolder){
    }
    protected internal virtual string MapPath(string virtualPath) {
        return System.Web.Hosting.HostingEnvironment.MapPath(virtualPath);
    }
    Dictionary<string, string> extensionsDisplayName;
    Dictionary<string, string> ExtensionsDisplayName {
        get {
            if(extensionsDisplayName == null)
                extensionsDisplayName = XDocument.Load(MapPath("~/Content/ExtensionsDisplayName.xml")).Descendants("Extension").ToDictionary(n => n.Attribute("Extension").Value, n => n.Attribute("DisplayName").Value);
            return extensionsDisplayName;
        }
    }
    public override IEnumerable<FileManagerFile> GetFiles(FileManagerFolder folder) {
        foreach(FileManagerFile file in base.GetFiles(folder)) {
            var fileInfo = new FileInfo(MapPath(file.FullName));
            FileManagerFileProperties fileProperties = new FileManagerFileProperties() {
                Metadata = new Dictionary<string, object>()
            };
            var fileSecurity = File.GetAccessControl(MapPath(file.FullName));
            var sid = fileSecurity.GetOwner(typeof(SecurityIdentifier));
            FileVersionInfo myFileVersionInfo = FileVersionInfo.GetVersionInfo(MapPath(file.FullName));
            var version = myFileVersionInfo.FileVersion;
            string description = myFileVersionInfo.FileDescription;
            fileProperties.Metadata.Add("Name", file.Name);
            fileProperties.Metadata.Add("Type", ExtensionsDisplayName[file.Extension]);
            fileProperties.Metadata.Add("CreationDate", fileInfo.CreationTime.ToString("U"));
            fileProperties.Metadata.Add("AccessedDate", fileInfo.LastAccessTime.ToString("U"));
            fileProperties.Metadata.Add("Attributes", fileInfo.Attributes);
            fileProperties.Metadata.Add("LastWriteTime", fileInfo.LastWriteTime.ToString("U"));
            fileProperties.Metadata.Add("Owner", sid.Value);
            fileProperties.Metadata.Add("Description", description);
            fileProperties.Metadata.Add("Size", fileInfo.Length.ToString("#,#") + " bytes");
            yield return new FileManagerFile(this, folder, file.Name, file.Id, fileProperties);
        }
    }
    public override IEnumerable<FileManagerFolder> GetFolders(FileManagerFolder parentFolder){
        foreach(FileManagerFolder folder in base.GetFolders(parentFolder)){
            var folderInfo = new FileInfo(MapPath(folder.FullName));
            FileManagerFolderProperties folderProperties = new FileManagerFolderProperties() {
                Metadata = new Dictionary<string, object>()
            };
            var folderSecurity = File.GetAccessControl(MapPath(folder.FullName));
            var sid = folderSecurity.GetOwner(typeof(SecurityIdentifier));
            folderProperties.Metadata.Add("Name", folder.Name);
            folderProperties.Metadata.Add("Type", "Folder");
            folderProperties.Metadata.Add("CreationDate", folderInfo.CreationTime.ToString("U"));
            folderProperties.Metadata.Add("AccessedDate", folderInfo.LastAccessTime.ToString("U"));
            folderProperties.Metadata.Add("Attributes", folderInfo.Attributes);
            folderProperties.Metadata.Add("LastWriteTime", folderInfo.LastWriteTime.ToString("U"));
            folderProperties.Metadata.Add("Owner", sid.Value);
            var directoryInfo = new DirectoryInfo(MapPath(folder.FullName));
            var filesInfo = directoryInfo.GetFiles("*", SearchOption.AllDirectories);
            long folderSize = 0;
            foreach(var info in filesInfo)
                folderSize += info.Length;
            var filesCount = filesInfo.Length;
            var subDirectoriesCount = directoryInfo.GetDirectories("*", SearchOption.AllDirectories).Length;
            folderProperties.Metadata.Add("Size", folderSize.ToString("#,#") + " bytes");
            folderProperties.Metadata.Add("Contains", string.Format("Files: {0}, Folders: {1}", filesCount, subDirectoriesCount));
            yield return new FileManagerFolder(this, folder, folder.Name, folder.Id, folderProperties);
        }
    }
}